lucene/core/src/java/org/apache/lucene/search/SliceExecutor.java

```diff
@@ -0,0 +1,105 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * Executor which is responsible
+ * for execution of slices based on the current status
+ * of the system and current system load
+ */
+class SliceExecutor {
+  private final Executor executor;
+
+  public SliceExecutor(Executor executor) {
+    this.executor = executor;
+  }
+
+  public <C> List<Future<C>> invokeAll(Collection<FutureTask<C>> tasks) {
+
+    if (tasks == null) {
+      throw new IllegalArgumentException("Tasks is null");
+    }
+
+    if (executor == null) {
+      throw new IllegalArgumentException("Executor is null");
+    }
+
+    List<Future<C>> futures = new ArrayList();
+
+    int i = 0;
+
+    for (FutureTask task : tasks) {
+      boolean shouldExecuteOnCallerThread = false;
+
+      // Execute last task on caller thread
+      if (i == tasks.size() - 1) {
+        shouldExecuteOnCallerThread = true;
+      }
+
+      processTask(task, futures, shouldExecuteOnCallerThread);
+      ++i;
+    }
+
+    return futures;
+  }
+
+  // Helper method to execute a single task
+  protected <C> void processTask(final FutureTask<C> task, final List<Future<C>> futures,
+                             final boolean shouldExecuteOnCallerThread) {
+    if (task == null) {
+      throw new IllegalArgumentException("Input is null");
+    }
+
+    if (!shouldExecuteOnCallerThread) {
+      try {
+        executor.execute(task);
+        futures.add(task);
+
+        return;
+      } catch (RejectedExecutionException e) {
+        // Execute on caller thread
+      }
+    }
+
+    runTaskOnCallerThread(task);
+
+    try {
+      futures.add(CompletableFuture.completedFuture(task.get()));
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  // Private helper method to run a task on the caller thread
+  private void runTaskOnCallerThread(FutureTask task) {
+    try {
+      task.run();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
```

Comment:
we don't need this catch block as task.run() doesn't declare any non-runtime exception?